fetch.js

```diff
@@ -68,58 +68,85 @@
     body.bodyUsed = true
   }
 
+  var charset_re = /;\s*charset=(\S+)/i
+
   function Body() {
     this._body = null
     this.bodyUsed = false
 
     this.arrayBuffer = function() {
-      throw new Error('Not implemented yet')
-    }
-
-    var blobSupport = (function() {
-      try {
-        new Blob();
-        return true
-      } catch(e) {
-        return false
+      var rejected = consumed(this)
+      if (rejected) {
+        return rejected
       }
-    })();
 
-    if (blobSupport) {
-      this.blob = function() {
-        var rejected = consumed(this)
-        return rejected ? rejected : Promise.resolve(new Blob([this._body]))
-      }
+      var body = this._body
+      return new Promise(function(resolve, reject) {
+        var fileReader = new FileReader()
+        fileReader.readAsArrayBuffer(body)
+        fileReader.onloadend = function(e) {
+          if (e.target.error) {
+            reject(e.target.error)
+          } else {
+            resolve(e.target.result)
+          }
+        }
+      })
+    }
+
+    this.blob = function() {
+      var rejected = consumed(this)
+      return rejected ? rejected : Promise.resolve(this._body)
     }
 
     if (self.FormData) {
       this.formData = function() {
-        var rejected = consumed(this)
-        return rejected ? rejected : Promise.resolve(decode(this._body))
+        return this.text().then(function(text) {
+          return new Promise(function(resolve, reject) {
+            try {
+              resolve(decode(text))
+            } catch (ex) {
+              reject(ex)
+            }
+          })
+        })
       }
     }
 
     this.json = function() {
+      return this.text().then(function(text) {
+        return new Promise(function(resolve, reject) {
+          try {
+            resolve(JSON.parse(text))
+          } catch (ex) {
+            reject(ex)
+          }
+        })
+      })
+    }
+
+    this.text = function() {
       var rejected = consumed(this)
       if (rejected) {
         return rejected
       }
 
+      var match = charset_re.exec(this.headers.get('Content-Type') || '')
+      var charset = match ? match[1] : null
       var body = this._body
       return new Promise(function(resolve, reject) {
-        try {
-          resolve(JSON.parse(body))
-        } catch (ex) {
-          reject(ex)
+        var fileReader = new FileReader()
+        fileReader.readAsText(body, charset)
```

Comment:
I think only utf-8 decoding `text()` will be supported.

> Resolve p with the result of running utf-8 decode on bytes.

https://fetch.spec.whatwg.org/#body-mixin